src/field_theory/splitting_field.lean

```diff
@@ -94,61 +95,405 @@ by rw [nat_degree_remove_factor, hfn, n.add_sub_cancel]
+  is_scalar_tower R₁ R₂ (adjoin_root f) :=
+⟨begin
+  intros a b x,
+  refine x.induction_on _ (λ p, _),
+  simp only [adjoin_root.smul_mk, smul_assoc a b p],
+end⟩
 
 instance is_scalar_tower (n : ℕ) : Π (R₁ R₂ : Type*) {K : Type u}
-  [comm_semiring R₁] [comm_semiring R₂] [has_smul R₁ R₂] [field K],
-  by exactI Π [algebra R₁ K] [algebra R₂ K],
-  by exactI Π [is_scalar_tower R₁ R₂ K] {f : K[X]},
-    is_scalar_tower R₁ R₂ (splitting_field_aux n f) :=
-nat.rec_on n (λ R₁ R₂ K _ _ _ _ _ _ _ _, by exactI ‹is_scalar_tower R₁ R₂ K›) $
-         λ n ih R₁ R₂ K _ _ _ _ _ _ _ f, by exactI ih R₁ R₂
+  [has_smul R₁ R₂] [field K],
+  by exactI Π [distrib_smul R₂ K] [distrib_smul R₁ K],
+  by exactI Π [is_scalar_tower R₁ K K] [is_scalar_tower R₂ K K] [is_scalar_tower R₁ R₂ K]
+    {f : K[X]}, by exactI is_scalar_tower R₁ R₂ (splitting_field_aux n f) :=
+nat.rec_on n (λ R₁ R₂ K _ _ hs₂ hs₁ _ _ h f,
+begin
+  rcases hs₁ with @⟨@⟨⟨hs₁⟩,_⟩,_⟩,
+  rcases hs₂ with @⟨@⟨⟨hs₂⟩,_⟩,_⟩,
+  exact h,
+end) $ λ n ih R₁ R₂ K _ _ _ _ _ _ _ f, by exactI ih R₁ R₂
+
+/-- Splitting fields have a negation. -/
+protected def neg (n : ℕ) : Π {K : Type u} [field K], by exactI Π {f : K[X]},
+  splitting_field_aux n f → splitting_field_aux n f :=
+nat.rec_on n (λ K fK f, by exactI @has_neg.neg K _) (λ n ih K fK f, ih)
+
+/-- Splitting fields have subtraction. -/
+protected def sub (n : ℕ) : Π {K : Type u} [field K], by exactI Π {f : K[X]},
+  splitting_field_aux n f → splitting_field_aux n f → splitting_field_aux n f :=
+nat.rec_on n (λ K fK f, by exactI @has_sub.sub K _) (λ n ih K fK f, ih)
+
+protected lemma nsmul_zero (n : ℕ) : Π {K : Type u} [field K], by exactI Π {f : K[X]}
+  (x : splitting_field_aux n f), (0 : ℕ) • x = splitting_field_aux.zero n :=
+nat.rec_on n (λ K fK f, by exactI zero_nsmul) (λ n ih K fK f, by exactI ih)
+
+protected lemma nsmul_succ (n : ℕ) : Π {K : Type u} [field K], by exactI Π {f : K[X]}
+  (k : ℕ) (x : splitting_field_aux n f),
+  (k + 1) • x = splitting_field_aux.add n x (k • x) :=
+nat.rec_on n (λ K fK f n x, by exactI succ_nsmul x n) (λ n ih K fK f, by exactI ih)
+
+protected lemma zsmul_zero (n : ℕ) : Π {K : Type u} [field K], by exactI Π {f : K[X]}
+  (x : splitting_field_aux n f), (0 : ℤ) • x = splitting_field_aux.zero n :=
+nat.rec_on n (λ K fK f, by exactI zero_zsmul) (λ n ih K fK f, by exactI ih)
+
+protected lemma zsmul_succ (n : ℕ) : Π {K : Type u} [field K], by exactI Π {f : K[X]}
+  (k : ℕ) (x : splitting_field_aux n f),
+  (↑(k + 1) : ℤ) • x = splitting_field_aux.add n x ((k : ℤ) • x) :=
+nat.rec_on n (λ K fK f k x, by exactI @subtraction_comm_monoid.zsmul_succ' K _ k x)
+  (λ n ih K fK f, by exactI ih)
+
+protected lemma zsmul_neg (n : ℕ) : Π {K : Type u} [field K], by exactI Π {f : K[X]}
+  (k : ℕ) (x : splitting_field_aux n f),
+  -[1+ k] • x = splitting_field_aux.neg n ((k + 1 : ℤ) • x) :=
+nat.rec_on n (λ K fK f k x, by exactI @subtraction_comm_monoid.zsmul_neg' K _ k x)
+  (λ n ih K fK f, by exactI ih)
```

Comment:
I think these should just be private, no one should be using them outside this file, right?